File: spatialsegregation/src/shannon.cpp
#include "shannon.h"



std::vector<double> shannon0(Graph *graph, double *fpar, int *dbg, int *included)
// this is obsolete: we compute only piitau:s in C and do the scaling in R
{
	if(*dbg)Rprintf("shannon[");
	int i,S=graph->pp->getNtypes();
	double Eglobal=0.0, Elocal=0.0, ratio, lambda0=0, X;
	std::vector<double> value;


	for(i=0;i<S;i++)// compute overall intensity&format piitau
		lambda0= lambda0 + graph->pp->lambdas[i];

	for(i=0;i<S;i++)// compute the global entropy
	{
		ratio = graph->pp->lambdas[i]/lambda0;
		X = 0.0;
		if(ratio>0) X =  graph->pp->lambdas[i] * (log((double)ratio) / log((double)S) );
		Eglobal= Eglobal + X;
	}
	Eglobal = - Eglobal/lambda0;//=~E(N)

	value = piitauf(graph, fpar, dbg, included); // compute the pi_tau's, most demanding part

	for(i=0;i<S;i++)//~pi_tau
	{
		ratio = value.at(i);
		if(ratio>0)
			Elocal = Elocal + ratio* (log((double)ratio)/log((double)S));
	}
	//and finally the estimate
	value.clear();
	value.push_back(0.0);

	value.at(0) =  -Elocal /(double) Eglobal;

	if(*dbg)Rprintf("]");
	return value;
}

std::vector<double> shannon(Graph *graph, double *fpar, int *dbg, int *included)
{
	if(fpar[0]==0) return piitauf(graph, fpar, dbg, included);
	else return shannon_v2(graph, fpar, dbg, included);
}


std::vector<double> piitauf(Graph *graph, double *fpar, int *dbg, int *included)
{
	if(*dbg)Rprintf("piitau[");
	int i,j,k,l,m, S=graph->pp->getNtypes();
	double piitauj, *piitau = new double[S];
	std::vector<double> value;
	value.clear();

	for(i=0;i<S;i++)//compute the pii_tau's
	{
//		if(graph->pp->lambdas[i]>0) // not right, should not exclude any targets
		{
			m = 0;
			piitau[i]=0.0;

			for(j=0;j< graph->pp->size();j++) // first the piitau for each node...
			 if(included[j]) // exclude the minus-sampling sources
			 {
				m=m+1;
				piitauj = 0.0;
				if(graph->nodelist[j].size()>0)
				{
					for(k=0;k<(int)graph->nodelist[j].size(); k++)
					{
						l = graph->nodelist[j][k]-1;
						if(graph->pp->getT(&l)==graph->pp->getTypevec(&i))
							piitauj = piitauj + 1;
					}
					piitauj = piitauj / (double)graph->nodelist[j].size();
				}
				piitau[i] = piitau[i] + piitauj; // sum for the arith. mean...
			 }
			if(m>0) piitau[i] = piitau[i]/(double) m;//then the arith. mean over all sampled nodes
			value.push_back(piitau[i]);
		}
	}
	if(*dbg)Rprintf("]");
	return value;
}


// Second version of local entropy: Use the real number of local species as the log-base,
// then calculate the mean of entropy per type. Returns per type values
std::vector<double> shannon_v2(Graph *graph, double *fpar, int *dbg, int *included)
{
	if(*dbg)Rprintf("local entropies[");
	int i,j,k, S=graph->pp->getNtypes(), *Htau = new int[S];
	double v, pii;
	std::vector<int> counts;
	std::vector<double> locpitau;
	std::vector<double> value;


	for(k=0;k<S;k++){counts.push_back(0);value.push_back(0.0);} // zero the counts and values

//	main loop over points
	for(i=0;i<graph->pp->size();i++)//compute the local entropy for each included node
	{
		if(included[i]) // focus point included in calculation
		{
			for(k=0; k< S; k++) Htau[k]=0; // set local type counts 0
			locpitau.clear();

			counts[graph->pp->getT(&i)-1]++; // add one to global count vector

			Htau[graph->pp->getT(&i)-1]++;   //TODO: Should the point be aware of itself's type?

			// add the counts of neighbour types to local count vector
			for(j=0; j< (int)graph->nodelist[i].size(); j++)
			{
				k = graph->nodelist[i][j]-1; // type index of neighbour j
				Htau[ graph->pp->getT(&k) ] = 1 + Htau[ graph->pp->getT(&k) ]; // add one to local count vector
			}

			// count which types are present
			for(k=0; k< S; k++)
			{
				if(Htau[k]>0) locpitau.push_back(Htau[k]);
			}

			v = 0;
			// do the local entropy calculation
			if(locpitau.size()>1) // if we have multitype neighbourhood (others than focal point type), otherwise entropy =0
				for(k=0;k < (int)locpitau.size();k++)
				{
					pii = locpitau[k]/(double)graph->nodelist[i].size(); // the local relative frequency of type k
					v = v + pii * (log((double)pii)/log((double)locpitau.size())); // sum over type pi*log(pi)
				}

			value.at(graph->pp->getT(&i)-1) = value.at(graph->pp->getT(&i)-1) + v; // contributes to its own type's mean entropy
		}
	}

	// now calculate the mean entropy per type
	for(k=0;k<S;k++)
	{
		if(counts.at(k)>0)
			value.at(k) = value.at(k)/(double)counts.at(k);
	}

	if(*dbg)Rprintf("]");
	return value;

}


//EOF

File: spatialsegregation/R/shannon.R
#' Spatial Shannon index
#' 
#' Compute the spatial and aspatial Shannon index for a given multitype point pattern.
#' 
#' @param X Multitype point pattern of class \code{ppp} (see package 'spatstat')
#' @param r Vector of sizes for neighbourhoods, e.g. \code{geometric} graph with different ranges.
#' @param spatial If FALSE, return the classical aspatial index value.
#' @param v2 If TRUE, use the real number of types in neighbourhoods as the log-base instead of total population type count.
#' @param ... Further parametes for the function \code{\link{segregationFun}}.
#' 
#' 
#' @details 
#' 
#' The form of Shannon index is  \var{H = 1 - E(o)/E(N)}, where \var{E(N)} is the global entropy and \var{E(o)} is the local entropy calculated as \var{E(o)= - sum pi_tau log(pi_tau)}, where the sum is over the different types present in the pattern, and \var{pi_tau} is the expected frequency of type \var{tau} points in a neighbourhood of a typical point of the pattern. 
#' 
#' The function \code{shannonF} is the calculation function. Uses function \code{\link{segregationFun}}. 
#' 
#' The function \code{shannon.index} is a shortcut to get the non-spatial Shannon index. 
#' 
#' @return 
#'  Returns an \code{fv}-object, see \code{spatstat} for more information. The index returns a scalar.
#'  
#' @references 
#' Rajala, Illian: A family of spatial biodiversity measures based on graphs, Env. Ecol. Stat. 2012
#'
#' Reardon, O'sullivan: Measures of spatial segregation. Sociological methodology, 34:121-162, 2004.
#' @export

shannonF<-function(X, r=NULL, v2=FALSE, ...)
#Shannon index for graphs, with possibly various range-parameters
{
	# check that X is multitype ppp-object
	verifyclass(X, "ppp")
	if(length(levels(X$marks))<2) stop("Use only on a multitype point pattern (data.frame-marks not yet supported).")
	
	# the main calc function: it calculates only the pi_tau-vector
	res<-segregationFun(X=X, r=r, fun="shannon", funpars=ifelse(v2,1,0), ...) 
	
	# calc the aspatial (global) entropy i.e. shannon index
	eglobal <- -shannon.index(X,spatial=FALSE)
	
	# a function to calculate  -1*sum( pi_tau*log(pi_tau) )
	f<-function(pvec) 
	{
		E1<-0
		pvec<-pvec/sum(pvec)
		S<-length(pvec)
		ok<-pvec>0
		E1 <- sum(pvec[ok]*log(pvec[ok],base=S))
		(1-E1/eglobal)
	}
	
	# if we take the log base as the individual degree instead of total S -
	#   then the result is the mean
	if(v2)
	{
		H<-apply(-res$v,1,mean)
		desc<-"Spatial Shannon index, v2 (check documentation)"
	}
	# the base is S: calc the E(o)/E
	else
	{
		H<-unname(apply(res$v,1,f))
		desc<-"Spatial Shannon index"
	}
		
	# the CSR values
	theo<-rep(0, length(res$parvec))
	
	# create the fv-object
	shannon.final<-fv(data.frame(theo=theo, par=res$parvec, H=H),
			          argu="par",
					  alim=range(res$parvec),
					  ylab=substitute(H,NULL),
					  desc=c("CSR values","Parameter values",desc),
					  valu="H",
					  fmla=".~par",
			      unitname=res$unitname,
					 fname=desc
					 )
	
	# add the typewise pii_tau values of which the index is a summary
	attr(shannon.final,"typewise")<- res$v
	
	# add the global index value too
	attr(shannon.final,"Aspatial Shannon index")<--eglobal
	
	# add the note about neighbourhood definition
	attr(shannon.final,"note")<-res$note
	
	# return
	shannon.final
}


####################################################################################################
#' @export 
#' @describeIn shannonF Traditional index.
shannon.index<-function(X, spatial=FALSE, ...)
{
	#the traditional aspatial Shannon index
	if(!spatial)
	{  
		sum0<-summary(X)
		ints<-sum0$marks[,3]
		m<-union(X$marks,NULL)
		pii<-ints/sum(ints)
		H<- (-1)*sum( pii * log(pii, length(m)),na.rm=TRUE)
		names(H)<-"Non-spatial Shannon index"
	}
	if(spatial)
	{			   #spatial Simpson index for a set of edgelists	
		H<-shannonF(X=X, ...)$H
		names(H)<-"Spatial Shannon index"
	}
	H
}
